### PR TITLE
Raise warnings on different wcs to WCSAxes

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/sunpy/sunpy/issues/5163.

This still needs:

- [x] Rolling out to other plotting methods
- [ ] Tests
- [ ] Changelog
- [x] A decision on backports

re. backporting, I think I am pro backporting a warning instead of an error, but interested in what others think.

### CODE REVIEW

File: sunpy/map/mapbase.py [hunk 1]
```diff
@@ -2258,7 +2258,7 @@ def plot(self, annotate=True, axes=None, title=True,
         if not axes:
             axes = wcsaxes_compat.gca_wcs(self.wcs)
 
-        if not wcsaxes_compat.is_wcsaxes(axes):
+        if not wcsaxes_compat.is_same_wcsaxes(axes, self.wcs):
             warnings.warn("WCSAxes not being used as the axes object for this plot."
```
Comment (Cadair) [inline sunpy/map/mapbase.py:2262]: This feels like the wrong warning for a mismatch? Also the mismatch should probably be an error in 3.0.

**Comment by Cadair** [comment]
+1 to backporting a warning.

**Comment by dstansby** [comment]
This is currently blocked by https://github.com/sunpy/sunpy/issues/5165

File: sunpy/map/mapbase.py [hunk 2]
```diff
@@ -2339,6 +2313,36 @@ def plot(self, annotate=True, axes=None, title=True,
...
+            if allow_non_wcsaxes:
+                warnings.warn("WCSAxes not being used as the axes object for this plot."
+                              " Plots may have unexpected behaviour. To fix this pass "
+                              "'projection=map' when creating the axes",
+                              SunpyUserWarning)
+            else:
+                raise TypeError("The axes need to be an instance of WCSAxes. You may have neglected "
+                                "to use `projection=` when creating the axes.")
+            # Check if the image is properly oriented
+            if not np.array_equal(self.rotation_matrix, np.identity(2)):
```
[the remark this answers is not in the document]
Comment (dstansby) [reply sunpy/map/mapbase.py:2337]: This is inside a non-wcsAxes block.

**Comment by dstansby** [comment]
I decided to have this as a warning, and then we can error in a later release.

**Comment by ayshih** [comment]
I'll point out there are use cases – that I'm promoting – of intentionally calling `draw_*` methods of a map with a different WCS than the WCSAxes (e.g., drawing the limb as seen from the observer of a different map, #5237).  I'd object to those situations always emitting warnings.

**Comment by dstansby** [comment]
> I'll point out there are use cases – that I'm promoting – of intentionally calling `draw_*` methods of a map with a different WCS than the WCSAxes (e.g., drawing the limb as seen from the observer of a different map, #5237). I'd object to those situations always emitting warnings.


If particular methods have (or gain) support for plotting on a WCSAxes with a different WCS, `allow_different_wcs=True` can currently be given to `_check_axes`, which I think is an appropriate resolution to this potential issue.

**Comment by ayshih** [comment]
Ah, I see.  So, of the six methods that use `_check_axes` in this PR:

- `draw_grid` -> different WCS is irrelevant, so it's already supported
- `draw_limb` -> different WCS will be supported in #5237 
- `draw_quadrangle` -> different WCS already supported
- `draw_rectangle` -> to be deprecated in #5236
- `contour` -> does it just need a coordinate transformation to support a different WCS?
- `plot` -> most effort to support a different WCS

I guess I see supporting a different WCS to be the norm rather than the outlier; I suggest changing the default.

**Comment by dstansby** [comment]
New PR (this was a hassle to rebase) here: https://github.com/sunpy/sunpy/pull/5244

**Pull request closed without merging**